codalib/xsdatetime.py

```diff
@@ -0,0 +1,188 @@
+from datetime import datetime, tzinfo, timedelta
+from pytz import timezone, utc as utc_tz
+
+# constants for time parsing/formatting
+# this is a stub
+XSDT_FMT = "%Y-%m-%dT%H:%M:%S"
+# this should never change
+XSDT_TZ_OFFSET = 19
+# the str value to use for default local timezone
+# in localization operations
+DEFAULT_LOCAL_TZ = "US/Central"
+
+
+class InvalidXSDateTime(Exception):
+    pass
+
+
+class XSDateTimezone(tzinfo):
+    """
+    Concrete subclass of tzinfo for making sense of timezone offsets.
+    Not really worried about localization here, just +/-HHMM
+    """
+
+    def __init__(self, hours=0, minutes=0, sign=1):
+        self.minutes = hours * 60 + minutes
+        self.minutes *= sign
+
+    def utcoffset(self, dt):
+        return timedelta(minutes=self.minutes)
+
+    def dst(self, dt):
+        return timedelta(0)
+
+
+def xsDateTime_parse(xdt_str, local_tz=None):
+    """
+    Parses xsDateTime strings of form 2017-01-27T14:58:00+0600, etc.
+    Returns a *naive* datetime in local time according to local_tz.
+    """
+    if not isinstance(xdt_str, basestring):
+        raise InvalidXSDateTime(
+            "Expecting str or unicode, got {}.".format(type(xdt_str))
+        )
+
+    try:
+        # this won't parse the offset (or other tzinfo)
+        naive_dt = datetime.strptime(xdt_str[0:XSDT_TZ_OFFSET], XSDT_FMT)
+    except:
+        raise InvalidXSDateTime("Malformed date/time ('%s')." % (xdt_str,))
+
+    naive_len = XSDT_TZ_OFFSET
+    offset_len = len(xdt_str) - naive_len
+    offset_str = xdt_str[-offset_len:]
+    offset_hours = None
+    offset_minutes = None
+    offset_sign = 1
+    parsed = None
+
+    # parse fractional seconds if present
+    fsec_i = 0
+    if not offset_len:
+        parsed = naive_dt
+    elif offset_str[0] is '.':
+        if offset_len > 1:
+            fsec_i = 1
+            fsec_chr = offset_str[fsec_i]
+            fsec = ''
+            while fsec_chr.isdigit():
+                fsec += fsec_chr
+                fsec_i += 1
+                if fsec_i >= offset_len:
+                    break
+                fsec_chr = offset_str[fsec_i]
+            fsec = float('.'+fsec)
+            naive_dt += timedelta(milliseconds=fsec*1000)
+        else:
+            raise InvalidXSDateTime('Malformed fractional seconds.')
+
+    # reset offset length and set offset string to actual offset,
+    # if we found fractional seconds -- otherwise this is all a noop
+    offset_len -= fsec_i
+    if offset_len:
+        offset_str = offset_str[fsec_i:fsec_i+offset_len+1]
+    else:
+        offset_str = ''
+
+    # get local timezone info using local_tz (str)
+    # throws pytz.exceptions.UnknownTimezoneError
+    # on bad timezone name
```

Comment:
Haven't brought up the PEP8 recommendation of capitalizing the first letter of comments previously because codalib already existed without respecting that and it is a minor thing anyway. Just going to mention here that one reason I like that recommendation is because it is easier to read comments that are on consecutive lines when some consist of one line and others span to multiple lines.